test/Test/Data/Map.purs

```diff
@@ -215,6 +215,12 @@ mapTests = do
           Just v              -> Just v == v2
           Nothing             -> not (in1 || in2)
 
+  log "difference"
+  quickCheck $ \(TestMap m1) (TestMap m2) ->
+    let d = M.difference m1 m2 :: M.Map SmallKey Int
+    in and (map (\k -> M.member k d) (A.fromFoldable $ M.keys d)) &&
```

Comment:
I think you meant to write `m1` instead of the last `d` on this line.

Reply:
Good catch. Fixed it. Want me to squash and force push or fine like this?

Reply:
Oh, sorry, that should indeed have been "first d".

I don't have the commit bit on this repo, but I'd assume someone would just hit the squash button.

I just added `difference` to -unordered-collections, fwiw.

Reply:
I had a look at the implementation in Haskell (below). There they obviously walk the tree, which I guess is far more efficient. Implementation in this PR follow the style of the other functions. Any thoughts on performance?

```haskell
difference :: Ord k => Map k a -> Map k b -> Map k a
difference Tip _   = Tip
difference t1 Tip  = t1
difference t1 (Bin _ k _ l2 r2) = case split k t1 of
  (l1, r1)
    | size l1l2 + size r1r2 == size t1 -> t1
    | otherwise -> link2 l1l2 r1r2
    where
      !l1l2 = difference l1 l2
      !r1r2 = difference r1 r2
```

Reply:
That would certainly be a lot faster. Repeatedly calling `delete` does a lot of duplicate work. Traversing the trees in parallel allows you to avoid retraversing parts of the tree while searching for elements to remove, avoid allocating upper nodes that will be replaced by a later deletion, and avoid looking for elements that can't possibly be there (like {1..100} \ {200..300}).

On the other hand, O(#right * log(#left)) is not terrible.

You probably know all of this.

So sure, it would be nice if it was faster, but IMHO that should not block merging.
...not that I have anything to say on the matter...